File: 6° Semestre/Python/AnalizadorLexico.py
import re

class Token:
    def __init__(self, tipo, valor):
        self.tipo = tipo
        self.valor = valor

class AnalizadorLexico:
    def __init__(self):
        self.patrones = [
            (r'[0-9]+', 'ENTERO'),
            (r'[a-zA-Z_][a-zA-Z0-9_]*', 'IDENTIFICADOR'),
            (r'\+', 'SUMA'),
            (r'-', 'RESTA'),
            (r'\*', 'MULTIPLICACION'),
            (r'/', 'DIVISION')
        ]

    def analizar(self, entrada):
        # Eliminar espacios en blanco
        entrada = entrada.replace(" ", "")
        tokens = []
        while entrada:
            for patron, etiqueta in self.patrones:
                match = re.match(patron, entrada)
                if match:
                    valor = match.group(0)
                    tokens.append(Token(etiqueta, valor))
                    entrada = entrada[len(valor):]
                    break
            else:
                raise ValueError(f"No se pudo analizar: {entrada}")
        return tokens

analizador = AnalizadorLexico()
entrada = input()
tokens = analizador.analizar(entrada)
for token in tokens:
    print(f"Tipo: {token.tipo}, Valor: {token.valor}")

File: 6° Semestre/Java/AnalizadorLexico.java
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class Token {
    String tipo;
    String valor;

    public Token(String tipo, String valor) {
        this.tipo = tipo;
        this.valor = valor;
    }
}

class AnalizadorLexico {
    List<String[]> patrones;

    public AnalizadorLexico() {
        patrones = new ArrayList<>();
        patrones.add(new String[] { "[0-9]+", "ENTERO" });
        patrones.add(new String[] { "[a-zA-Z_][a-zA-Z0-9_]*", "IDENTIFICADOR" });
        patrones.add(new String[] { "\\+", "SUMA" });
        patrones.add(new String[] { "-", "RESTA" });
        patrones.add(new String[] { "\\*", "MULTIPLICACION" });
        patrones.add(new String[] { "/", "DIVISION" });
    }

    public List<Token> analizar(String entrada) {
        // Eliminar espacios en blanco
        entrada = entrada.replace(" ", "");
        List<Token> tokens = new ArrayList<>();
        while (!entrada.isEmpty()) {
            for (String[] patron : patrones) {
                Pattern pattern = Pattern.compile(patron[0]);
                Matcher matcher = pattern.matcher(entrada);
                if (matcher.find() && matcher.start() == 0) {
                    String valor = matcher.group();
                    tokens.add(new Token(patron[1], valor));
                    entrada = entrada.substring(valor.length());
                    break;
                }
            }
        }
        return tokens;
    }

    public static void main(String[] args) {
        AnalizadorLexico analizador = new AnalizadorLexico();
        String entrada = "3 + 4 * x - y / 2";
        List<Token> tokens = analizador.analizar(entrada);
        for (Token token : tokens) {
            System.out.println("Tipo: " + token.tipo + ", Valor: " + token.valor);
        }
    }
}
